File: Samples for Parallel Programming with the .NET Framework/[C#][VB][C++][F#]-Samples for Parallel Programming with the .NET Framework/C# and C++ and F# and VB/C#,C++,F#,VB/Sudoku/Sudoku_CSharp/SolverGenerator/Techniques/BlockAndColumnRowInteractionTechnique.cs
//--------------------------------------------------------------------------
// 
//  Copyright (c) Microsoft Corporation.  All rights reserved. 
// 
//  File: BlockAndColumnRowInteractionTechnique.cs
//
//  Description: Implements the Sudoku block and column/row interaction technique.
// 
//--------------------------------------------------------------------------

using System;
using System.Drawing;
using Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples.Sudoku.Collections;

namespace Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples.Sudoku.Techniques
{
	/// <summary>Implements the coloring elimination technique.</summary>
	[Serializable]
    public sealed class BlockAndColumnRowInteractionTechnique : EliminationTechnique
	{
		/// <summary>Initialize the technique.</summary>
		public BlockAndColumnRowInteractionTechnique(){}

		/// <summary>Gets the difficulty level of this technique.</summary>
		internal override uint DifficultyLevel { get { return 4; } }

        public override bool Equals(object obj)
        {
            return obj is BlockAndColumnRowInteractionTechnique;
        }

        public override int GetHashCode()
        {
            return GetType().GetHashCode();
        }

		/// <summary>Runs this elimination technique over the supplied puzzle state and previously computed possible numbers.</summary>
		/// <param name="state">The puzzle state.</param>
		/// <param name="possibleNumbers">The previously computed possible numbers.</param>
		/// <param name="numberOfChanges">The number of changes made by this elimination technique.</param>
		/// <returns>Whether more changes may be possible based on changes made during this execution.</returns>
		internal override bool Execute(
			PuzzleState state, bool exitEarlyWhenSoleFound, 
			FastBitArray[][] possibleNumbers, out int numberOfChanges, out bool exitedEarly)
		{
			numberOfChanges = 0;
			exitedEarly = false;

			// Find the open cells in this block
			int numLocations;
			Point [] foundLocations = new Point[state.GridSize];

			// Analyze each box
			for(int box=0; box<state.GridSize; box++)
			{
				for(int n=0; n<state.GridSize; n++)
				{
					numLocations = 0;

					// Find all locations that n is possible within the box
					int boxStartX = (box / state.BoxSize) * state.BoxSize;
					for (int x = boxStartX; x < boxStartX + state.BoxSize && numLocations <= state.BoxSize; x++)
					{
						int boxStartY = (box % state.BoxSize) * state.BoxSize;
						for (int y = boxStartY; y < boxStartY + state.BoxSize && numLocations <= state.BoxSize; y++)
						{
							if (possibleNumbers[x][y][n])
							{
								foundLocations[numLocations++] = new Point(x,y);
							}
						}
					}

					// We only care if two or three are open in the grid and if they're
					// in the same row or column
					if (numLocations > 1 && numLocations <= state.BoxSize)
					{
						bool matchesRow = true, matchesColumn = true;
						int row = foundLocations[0].X;
						int column = foundLocations[0].Y;
						for(int i=1; i<numLocations; i++)
						{
							if (foundLocations[i].X != row) matchesRow = false;
							if (foundLocations[i].Y != column) matchesColumn = false;
						}

						// If they're all in the same row
						if (matchesRow)
						{
							for(int j=0; j<state.GridSize; j++)
							{
								if (possibleNumbers[row][j][n] && 
									j != foundLocations[0].Y && 
									j != foundLocations[1].Y && 
									(numLocations == 2 || j != foundLocations[2].Y)) // only works if BoxSize == 3
								{
									possibleNumbers[row][j][n] = false;
									numberOfChanges++;
									if (exitEarlyWhenSoleFound &&
										possibleNumbers[row][j].CountSet == 1)
									{
										exitedEarly = true;
										return false;
									}
								}
							}
						}
							// If they're all in the same column
						else if (matchesColumn)
						{
							for(int j=0; j<state.GridSize; j++)
							{
								if (possibleNumbers[j][column][n] &&
									j != foundLocations[0].X && 
									j != foundLocations[1].X && 
									(numLocations == 2 || j != foundLocations[2].X)) // only works if BoxSize == 3
								{
									possibleNumbers[j][column][n] = false;
									numberOfChanges++;
									if (exitEarlyWhenSoleFound &&
										possibleNumbers[j][column].CountSet == 1)
									{
										exitedEarly = true;
										return false;
									}
								}
							}
						}
					}
				}
			}

			return numberOfChanges != 0;
		}
	}
}

File: Samples for Parallel Programming with the .NET Framework/[C#][VB][C++][F#]-Samples for Parallel Programming with the .NET Framework/C# and C++ and F# and VB/C#,C++,F#,VB/Sudoku/Sudoku_VisualBasic/SolverGenerator/Techniques/BlockAndColumnRowInteractionTechnique.vb
'--------------------------------------------------------------------------
' 
'  Copyright (c) Microsoft Corporation.  All rights reserved. 
' 
'  File: BlockAndColumnRowInteractionTechnique.vb
'
'  Description: Implements the Sudoku block and column/row interaction technique.
' 
'--------------------------------------------------------------------------

Imports Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples.Sudoku.Collections

Namespace Microsoft.ParallelComputingPlatform.ParallelExtensions.Samples.Sudoku.Techniques
	''' <summary>Implements the coloring elimination technique.</summary>
	<Serializable>
	Public NotInheritable Class BlockAndColumnRowInteractionTechnique
		Inherits EliminationTechnique
		''' <summary>Initialize the technique.</summary>
		Public Sub New()
		End Sub

		''' <summary>Gets the difficulty level of this technique.</summary>
		Friend Overrides ReadOnly Property DifficultyLevel() As UInteger
			Get
				Return 4
			End Get
		End Property

		Public Overrides Overloads Function Equals(ByVal obj As Object) As Boolean
			Return TypeOf obj Is BlockAndColumnRowInteractionTechnique
		End Function

		Public Overrides Function GetHashCode() As Integer
			Return Me.GetType().GetHashCode()
		End Function

		''' <summary>Runs this elimination technique over the supplied puzzle state and previously computed possible numbers.</summary>
		''' <param name="state">The puzzle state.</param>
		''' <param name="possibleNumbers">The previously computed possible numbers.</param>
		''' <param name="numberOfChanges">The number of changes made by this elimination technique.</param>
		''' <returns>Whether more changes may be possible based on changes made during this execution.</returns>
        Friend Overrides Function Execute(ByVal state As PuzzleState, ByVal exitEarlyWhenSoleFound As Boolean,
                ByVal possibleNumbers()() As FastBitArray, <System.Runtime.InteropServices.Out()> ByRef numberOfChanges As Integer,
                <System.Runtime.InteropServices.Out()> ByRef exitedEarly As Boolean) As Boolean
            numberOfChanges = 0
            exitedEarly = False

            ' Find the open cells in this block
            Dim numLocations As Integer
            Dim foundLocations(state.GridSize - 1) As Point

            ' Analyze each box
            For box = 0 To state.GridSize - 1
                For n = 0 To state.GridSize - 1
                    numLocations = 0

                    ' Find all locations that n is possible within the box
                    Dim boxStartX = (box \ state.BoxSize) * state.BoxSize
                    Dim x = boxStartX
                    Do While x < boxStartX + state.BoxSize AndAlso numLocations <= state.BoxSize
                        Dim boxStartY = (box Mod state.BoxSize) * state.BoxSize
                        Dim y = boxStartY
                        Do While y < boxStartY + state.BoxSize AndAlso numLocations <= state.BoxSize
                            If possibleNumbers(x)(y)(n) Then
                                foundLocations(numLocations) = New Point(x, y)
                                numLocations += 1
                            End If
                            y += 1
                        Loop
                        x += 1
                    Loop

                    ' We only care if two or three are open in the grid and if they're
                    ' in the same row or column
                    If numLocations > 1 AndAlso numLocations <= state.BoxSize Then
                        Dim matchesRow = True, matchesColumn = True
                        Dim row = foundLocations(0).X
                        Dim column = foundLocations(0).Y
                        For i = 1 To numLocations - 1
                            If foundLocations(i).X <> row Then
                                matchesRow = False
                            End If
                            If foundLocations(i).Y <> column Then
                                matchesColumn = False
                            End If
                        Next i

                        ' If they're all in the same row
                        If matchesRow Then
                            For j = 0 To state.GridSize - 1
                                If possibleNumbers(row)(j)(n) AndAlso j <> foundLocations(0).Y AndAlso j <>
                                    foundLocations(1).Y AndAlso (numLocations = 2 OrElse j <> foundLocations(2).Y) Then ' only works if BoxSize == 3
                                    possibleNumbers(row)(j)(n) = False
                                    numberOfChanges += 1
                                    If exitEarlyWhenSoleFound AndAlso possibleNumbers(row)(j).CountSet = 1 Then
                                        exitedEarly = True
                                        Return False
                                    End If
                                End If
                            Next j
                            ' If they're all in the same column
                        ElseIf matchesColumn Then
                            For j = 0 To state.GridSize - 1
                                If possibleNumbers(j)(column)(n) AndAlso j <> foundLocations(0).X AndAlso j <>
                                    foundLocations(1).X AndAlso (numLocations = 2 OrElse j <> foundLocations(2).X) Then ' only works if BoxSize == 3
                                    possibleNumbers(j)(column)(n) = False
                                    numberOfChanges += 1
                                    If exitEarlyWhenSoleFound AndAlso possibleNumbers(j)(column).CountSet = 1 Then
                                        exitedEarly = True
                                        Return False
                                    End If
                                End If
                            Next j
                        End If
                    End If
                Next n
            Next box

            Return numberOfChanges <> 0
        End Function
	End Class
End Namespace